c/cert/src/rules/EXP36-C/DoNotCastPointerToMoreStrictlyAlignedPointerType.ql

```diff
@@ -0,0 +1,170 @@
+import semmle.code.cpp.dataflow.DataFlow
+import semmle.code.cpp.rangeanalysis.SimpleRangeAnalysis
+import DataFlow::PathGraph
+
+/**
+ * An expression with a type that has defined alignment requirements
+ */
+abstract class ExprWithAlignment extends Expr {
+  /**
+   * Gets the alignment requirements in bytes for the underlying `Expr`
+   */
+  abstract int getAlignment();
+
+  /**
+   * Gets a descriptive string describing the type of expression
+   */
+  abstract string getKind();
+}
+
+/**
+ * A class extending `AddressOfExpr` and `ExprWithAlignment` to reason about the
+ * alignment of base types addressed with C address-of expressions
+ */
+class AddressOfAlignedVariableExpr extends AddressOfExpr, ExprWithAlignment {
+  AddressOfAlignedVariableExpr() { this.getAddressable() instanceof Variable }
+
+  AlignAs alignAsAttribute() { result = this.getAddressable().(Variable).getAnAttribute() }
+
+  override int getAlignment() {
+    result = alignAsAttribute().getArgument(0).getValueInt()
+    or
+    result = alignAsAttribute().getArgument(0).getValueType().getSize()
+    or
+    not exists(alignAsAttribute()) and
+    result = this.getAddressable().(Variable).getType().getAlignment()
+  }
+
+  override string getKind() { result = "address-of expression" }
+}
+
+/**
+ * A class extending `FunctionCall` and `ExprWithAlignment` to reason about the
+ * alignment of pointers allocated with calls to C standard library allocation functions
+ */
+class DefinedAlignmentAllocationExpr extends FunctionCall, ExprWithAlignment {
+  int alignment;
+
+  DefinedAlignmentAllocationExpr() {
+    this.getTarget().getName() = "aligned_alloc" and
+    lowerBound(this.getArgument(0)) = upperBound(this.getArgument(0)) and
+    alignment = upperBound(this.getArgument(0))
+    or
+    this.getTarget().getName() = ["malloc", "calloc", "realloc"] and
+    alignment = getGlobalMaxAlignT()
+  }
+
+  override int getAlignment() { result = alignment }
+
+  override string getKind() { result = "call to " + this.getTarget().getName() }
+}
+
+/**
+ * A class extending `VariableAccess` and `ExprWithAlignment` to reason about the
+ * alignment of pointers accessed based solely on the pointers' base types.
+ */
+class DefaultAlignedPointerAccessExpr extends VariableAccess, ExprWithAlignment {
+  DefaultAlignedPointerAccessExpr() {
+    this.getTarget().getUnspecifiedType() instanceof PointerType and
+    not this.getTarget().getUnspecifiedType() instanceof VoidPointerType
+  }
+
+  override int getAlignment() {
+    result = this.getTarget().getType().(PointerType).getBaseType().getAlignment()
+  }
+
+  override string getKind() { result = "pointer base type" }
+}
+
+/**
+ * A data-flow configuration for analysing the flow of `ExprWithAlignment` pointer expressions
+ * to casts which perform pointer type conversions and potentially create pointer alignment issues.
+ */
+class ExprWithAlignmentToCStyleCastConfiguration extends DataFlow::Configuration {
```

Comment:
I've found some false positives on OpenPilot that are related to overly permissive dataflow. Essentially, the flow is as follows:
```
char c = 'c';
char * cp = &c;
char* a[3];
a[0] = cp; // Store pointer in array of pointers
char ** cpp = a; // An array of pointers is a char** so this is 
```

The problem is that global dataflow tracks through `a[0] = cp;` to subsequent reads of `a` (not just `a[0]`). I think this could be fixed with a sanitizer on the dataflow configuration.

Reply:
Changes made.